Remove `is_stateless` flag

hwwhww opened the pull request:
### What was wrong?
Fix #268 

### How was it fixed?
- [x] Removed `is_stateless` flag, only use the new route in `VM.apply_transaction`.

#### Cute Animal Picture

![koala-630117_1280](https://user-images.githubusercontent.com/9263930/35090626-2eb5679a-fc75-11e7-9d49-3c20ed74bb71.jpg)

pipermerriam commented on evm/vm_state.py:
```diff
@@ -179,35 +179,27 @@ def get_computation(self, message):
-            block.header.state_root = self.state_root
-            return computation, None, None
+        # Don't modify the given block
+        block = copy.deepcopy(block)
```
What do you think about removing this `deepcopy` and instead making `block` immutable and modifying the logic inside of `add_transaction` that creates and returns a new block object.

hwwhww commented:
#### Changelogs
1. Updated Chain.header in Chain.apply_transaction + Added `chaindb.persist_trie_data_dict_to_db`
2. Create a pure new block in `VMState.add_transaction`

I added `BlockHeader.pure_clone(header)` class method to do the dumb copy:
[py-evm/headers.py at 8f4aab7eba7a17e9491c7bb49282a5ba7261670b · hwwhww/py-evm · GitHub](https://github.com/hwwhww/py-evm/blob/8f4aab7eba7a17e9491c7bb49282a5ba7261670b/evm/rlp/headers.py#L137-L155)

* An alternative solution is do the `copy.deepcopy(header)` in `VMState.add_transaction` and then call `rlp.make_mutable`, but it’s against the goal of removing `deepcopy`.
* I'm not sure whether the naming of `pure_clone` is appropriate, it's just my instinct.

hwwhww commented on evm/db/chain.py:
```diff
@@ -316,6 +316,13 @@ def persist_block_to_db(self, block):
             rlp.encode(block.uncles, sedes=rlp.sedes.CountableList(type(block.header))),
         )
 
+    def persist_trie_data_dict_to_db(self, trie_data_dict):
```
But it _does_ use `self.db`...?

hwwhww replied:
But it _does_ use `self.db`...?

pipermerriam replied:
Disregard, I missed it somehow

hwwhww commented on evm/rlp/headers.py:
```diff
@@ -133,3 +133,23 @@ def from_parent(cls,
+
+    @classmethod
+    def pure_clone(cls, header):
+        return cls(
```
Your way looks much better for maintaining the possible new fields, thank you!

hwwhww commented:
@pipermerriam Thank you for the review feedback, I added the patch commit.
